onnxruntime/contrib_ops/rocm/bert/fast_gelu_impl.cu

```diff
@@ -31,91 +31,34 @@ limitations under the License.
-        }
-      } else {
-        if (0 == (input_length % 8) && (input_length >= 3145728)) {  // 3145728=8*128*3072
-          const int grid_size = (input_length / 8 + block_size - 1) / block_size;
-          hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluKernelVec<half, block_size, 8>), dim3(grid_size),
-                                             dim3(block_size), 0, stream, input_length, bias_length,
-                                             input, bias, output);
-        } else if (0 == (input_length % 4)) {
-          const int grid_size = (input_length / 4 + block_size - 1) / block_size;
-          hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluKernelVec<half, block_size, 4>), dim3(grid_size),
-                                             dim3(block_size), 0, stream, input_length, bias_length,
-                                             input, bias, output);
-        } else if (0 == (input_length % 2)) {
-          const int grid_size = (input_length / 2 + block_size - 1) / block_size;
-          hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluKernelVec<half, block_size, 2>), dim3(grid_size),
-                                             dim3(block_size), 0, stream, input_length, bias_length,
-                                             input, bias, output);
-        } else {
-          const int grid_size = (input_length + block_size - 1) / block_size;
-          hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluKernel<half, block_size>), dim3(grid_size),
-                                             dim3(block_size), 0, stream, input_length, bias_length,
-                                             input, bias, output);
-        }
-      }
-  } else {
-    const int grid_size = (input_length + block_size - 1) / block_size;
-    hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluKernel<half, block_size>), dim3(grid_size),
-                                       dim3(block_size), 0, stream, input_length, bias_length,
-                                       input, bias, output);
+                          const T* input, const T* bias, T* output, bool tuning) {
+  static FastGeluTunableOp<T> op;
+  if (tuning) {
+    op.EnableTuning();
   }
+  FastGeluParams<T> op_params(stream, input, bias, output, input_length, bias_length);
+  op.Run(&op_params);
   return HIP_CALL(hipPeekAtLastError());
 }
 
-template <>
-bool LaunchFastGeluKernel(hipStream_t stream, int input_length, int bias_length,
-                          const BFloat16* input, const BFloat16* bias, BFloat16* output, bool /*use_half2*/) {
-  constexpr int block_size = 256;
-  const int grid_size = (input_length + block_size - 1) / block_size;
-  hipLaunchKernelGGL(HIP_KERNEL_NAME(FastGeluKernel<BFloat16, block_size>), dim3(grid_size), dim3(block_size), 0,
-                     stream, input_length, bias_length, input, bias, output);
-  return HIP_CALL(hipPeekAtLastError());
-}
+template bool LaunchFastGeluKernel<float>(hipStream_t stream, int input_length, int bias_length,
```

Comment:
Out of curiosity, why do we exercise explicit instantiation here, instread of moving the template to a header file?

Reply:
This is to follow the convention of existing structure and I think the reason of explicit instantiation to seperate kernel code compilation from host code (to avoid putting it in header file and having it included by the host code).